### PR TITLE
Include support for the scatter gather cuda kernels to allow for comp…

### LINKED ISSUE #121965: scatter_reduce method do not support complex number multiplication on CUDA
### 🐛 Describe the bug

```python
import torch


device = torch.device('cuda:0') if torch.cuda.is_available() else torch.device("cpu")


# Complex tensors
src = torch.tensor([1 + 2j, 3 - 4j, 1 + 1j], dtype=torch.cfloat, device=device)
indices = torch.tensor([0, 0, 0], device=device)

# src = torch.tensor([1 + 2j, 3 - 4j, 1 + 1j], dtype=torch.complex64)
# indices = torch.tensor([0, 0, 0])


output = src.scatter_reduce(dim=0, index=indices, src=src, reduce='prod', include_self=False)
print(output)
```

Error message:
```
Traceback (most recent call last):
  File "F:\new_develop\spatea\test\torch_scatter_complex\test.py", line 15, in <module>
    output = src.scatter_reduce(dim=0, index=indices, src=src, reduce='prod', include_self=False)
RuntimeError: "cuda_scatter_gather_base_kernel_func" not implemented for 'ComplexFloat'
```

However, this issue will not happen for CPU device. This seems related to issue [74770](https://github.com/pytorch/pytorch/issues/74770), and [test_code](https://github.com/pytorch/pytorch/blob/4cbf963894e78d1cfedffe4f829740dc99163caa/test/test_scatter_gather_ops.py#L182)

### Versions

2.0.0+cu117

cc @ptrblck @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @amjames

### PR DESCRIPTION
**Pull request by ZelboK**
Fixes #121965 

This PR hopes to add support complex numbers in the scatter/gather related kernels. For brevity, I will only include `complex<float>` for now as `complex<double>`, for example, will be more complicated. 

C++ unit tests are currently passing alongside tests in `test_scatter_gather_ops.py`. Python test suites also seem to be passing. 

Please keep the following in mind:
1) I think this is my first time using Pytorch.
2) This is my first contribution to Pytorch. 

Environment:
3080 & WSL 2. `nvcc` is at 12.4. 

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

### CODE REVIEW

File: aten/src/ATen/cuda/Atomic.cuh
Comment (ZelboK) [inline aten/src/ATen/cuda/Atomic.cuh:434]: Inlined to avoid multiple definition errors.

Using the magnitude to order the complex numbers(since they are not straight forward to order). I am using intrinsic functions from CUDA but please let me know if these functions should also be callable from host code - if so, I can resort to regular instructions.

File: aten/src/ATen/cuda/Atomic.cuh
Comment (ZelboK) [inline aten/src/ATen/cuda/Atomic.cuh:54]: So `complex<float>` is more straight forward than `complex<double>`. I am leaving it as this implementation for now. To implement `complex<double>`, one avenue that could be taken is to use locking for example but there will be performance costs. 

However, `atomicCAS` does support 128 bits on the hopper architecture.

**Comment by janeyx99** [comment]
Requesting for review from @mikaylagawarecki who's worked with scatter_gather and also from @ptrblck regarding the cuda side.

File: aten/src/ATen/native/cuda/ScatterGatherKernel.cu
Comment (eqy) [inline aten/src/ATen/native/cuda/ScatterGatherKernel.cu:7]: nit: cleanup

**Review by eqy — DISMISSED** [review]

File: aten/src/ATen/native/cuda/ScatterGatherKernel.cu
Comment (ZelboK) [reply aten/src/ATen/native/cuda/ScatterGatherKernel.cu:7]: Oh wow. I missed this, sorry!

**Comment by ZelboK** [comment]
@mikaylagawarecki  @eqy I see that the pipeline is failing because of a linting issue. I used the lintrunner - I do not believe I touched this line on my own. https://github.com/pytorch/pytorch/actions/runs/8824920734/job/24228735042#step:11:244

I can change it if need be

File: test/test_torch.py
Comment (ZelboK) [reply test/test_torch.py]: Yeah, let me push the fix.

```
...
ATEN_CPU_GPU_AGNOSTIC success!
RUFF success!
LINTRUNNER_VERSION success!
ok No lint issues.
Successfully applied all patches.
```
Lint runner is happy locally atm

**Comment by ZelboK** [comment]
@eqy Anything that needs to be done on my end?

**Comment by ZelboK** [comment]
@eqy @mikaylagawarecki 

Hi folks, since I'm new to Pytorch I'm curious to know what the procedure is now. Will this be reviewed by a core maintainer/contributor by way of triage? Do I need to do anything on my end?

**Review by mikaylagawarecki — DISMISSED** [review]

**Comment by mikaylagawarecki** [comment]
Sorry about that :( @Zelbolk could you address the failure please

**Comment by ZelboK** [comment]
> @pytorchbot revert -m "windows build failure is real, https://github.com/pytorch/pytorch/actions/runs/8910674030/job/24470387612#step:11:11236 is the correct failure line, ignore the statement saying build passed, batch is errorcodes arent propagating again" -c ignoredsignal

The workflow is kind of misleading. Would it be helpful to file an issue? 

@mikaylagawarecki My apologies for missing this! I will push a fix now.

**Comment by clee2000** [comment]
> @clee2000 The workflow is kind of misleading. Would it be helpful to file an issue?

It should be fixed by https://github.com/pytorch/pytorch/pull/125306

**Review by mikaylagawarecki — DISMISSED** [review]
Looks like the failure is resolved now, so stamping to allow merge

**Comment by kit1980** [comment]
I'm going to revert this at multiple internal builds failed with things like

```
fbcode/caffe2/aten/src/ATen/cuda/Atomic.cuh(374): error: calling a constexpr __host__ function("operator*=") from a __device__ function("operator()") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.

fbcode/caffe2/aten/src/ATen/cuda/Atomic.cuh(450): error: calling a constexpr __host__ function("max") from a __device__ function("complex_max") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
```

@mikaylagawarecki see D56861849 if you want to help re-landing this.

**Comment by ZelboK** [comment]
@mikaylagawarecki Really sorry for the trouble :( 
> I'm going to revert this at multiple internal builds failed with things like
> 
> ```
> fbcode/caffe2/aten/src/ATen/cuda/Atomic.cuh(374): error: calling a constexpr __host__ function("operator*=") from a __device__ function("operator()") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
> 
> fbcode/caffe2/aten/src/ATen/cuda/Atomic.cuh(450): error: calling a constexpr __host__ function("max") from a __device__ function("complex_max") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
> ```
> 
> @mikaylagawarecki see D56861849 if you want to help re-landing this.

Is there a way for me to replicate one of these environments? Hopefully a devcontainer?

File: aten/src/ATen/cuda/Atomic.cuh
Comment (mikaylagawarecki) [inline aten/src/ATen/cuda/Atomic.cuh:371]: Looking at the failure and other function definitions, it looks like we need `__host__ __device__` here and for `complex_max` on 433, I can import and see whether this fixes the internal build tomorrow morning

**Review by eqy — APPROVED** [review]

**Comment by mikaylagawarecki** [comment]
@ZelboK I think the *= update in complex.h you mentioned is still needed, as the internal build still has errors like such

```
fbcode/caffe2/aten/src/ATen/cuda/Atomic.cuh(374): error: calling a constexpr __host__ function("operator*=") from a __device__ function("operator()") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
```

Also you seem to have accidentally commited ideep and TensorBase.cpp in this PR 😅 

I can import this again to check after you make these fixes, so do let me know when! 

fyi: In case you don't get a response from me, wanted to let you know that I will be out this coming week but will be back on Monday (5/13)

**Comment by ZelboK** [comment]
> @ZelboK I think the *= update in complex.h you mentioned is still needed, as the internal build still has errors like such
> 
> ```
> fbcode/caffe2/aten/src/ATen/cuda/Atomic.cuh(374): error: calling a constexpr __host__ function("operator*=") from a __device__ function("operator()") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
> ```
> 
> Also you seem to have accidentally commited ideep and TensorBase.cpp in this PR 😅
> 
> I can import this again to check after you make these fixes, so do let me know when!
> 
> fyi: In case you don't get a response from me, wanted to let you know that I will be out this coming week but will be back on Monday (5/13)

😂 my bad, didn't mean to commit that. 

 All good, thanks for the prompt assistance! I enjoyed working through this, including the build failures and all. I appreciate the helpfulness from the team ❤️ 

I'll push

File: aten/src/ATen/cuda/Atomic.cuh
Comment (peterbell10) [inline aten/src/ATen/cuda/Atomic.cuh:54]: This isn't atomic? You need to return `csum` directly, otherwise the value at `addr_as_ull` may change underneath you.

File: aten/src/ATen/cuda/Atomic.cuh
Comment (peterbell10) [inline aten/src/ATen/cuda/Atomic.cuh:450]: Is there any precedence for this definition of complex max/min in PyTorch?

**Review by peterbell10 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/cuda/Atomic.cuh
Comment (peterbell10) [reply aten/src/ATen/cuda/Atomic.cuh:54]: In fact this is also wrong as atomic read-modify-write ops return the old value, not the new value. So this should be bit-casting `assumed`.

File: aten/src/ATen/cuda/Atomic.cuh
Comment (ZelboK) [reply aten/src/ATen/cuda/Atomic.cuh:450]: I am not experienced enough with Pytorch to answer that. Aside from using magnitudes how else would you order them?

File: aten/src/ATen/cuda/Atomic.cuh
Comment (ZelboK) [reply aten/src/ATen/cuda/Atomic.cuh:54]: > In fact this is also wrong as atomic read-modify-write ops return the old value, not the new value. So this should be bit-casting `assumed`.

Sorry for the oversight. Could you help me understand? I know that `atomicCAS` returns the old value but with what in mind are you referring that to?

File: aten/src/ATen/cuda/Atomic.cuh
Comment (lezcano) [reply aten/src/ATen/cuda/Atomic.cuh:450]: Exactly, they cannot be ordered (in mathematical terms, complex numbers are not an [ordered field](https://en.wikipedia.org/wiki/Ordered_field)) 
We should error in these cases, same as we error when we call `max` on a complex tensor. If people want to use these ops on complex tensors, they can do a `view_as_real` and perform some transformations on the output to define the order they want.

File: aten/src/ATen/cuda/Atomic.cuh
Comment (peterbell10) [reply aten/src/ATen/cuda/Atomic.cuh:54]: `assumed` is the value before performing the update, which is what is returned by normal `atomicAdd`, `atomicMax`, etc.

See the CAS implementation for half as an example:
https://github.com/pytorch/pytorch/blob/6bdcc8d58867ed7a87e53939d502f667a9869bf2/aten/src/ATen/cuda/Atomic.cuh#L33-L34

File: aten/src/ATen/cuda/Atomic.cuh
Comment (ZelboK) [reply aten/src/ATen/cuda/Atomic.cuh:450]: I was under the impression that some contexts will use magnitude for ordering complex numbers, like spectral analysis for DSP. I also took motivation from https://www.mathworks.com/help/matlab/ref/max.html as well

@Franklalalala 

Could you comment on whether or not you had a use case for scattering complex numbers? What kind of work were you trying to do? Would you know if ordering of complex numbers is practically useful?

File: aten/src/ATen/cuda/Atomic.cuh
Comment (ZelboK) [reply aten/src/ATen/cuda/Atomic.cuh:54]: Oh you're right. I got tunnel visioned on the line on the actual call of `atomicCAS`, yes it should be `assumed`.

File: aten/src/ATen/cuda/Atomic.cuh
Comment (lezcano) [reply aten/src/ATen/cuda/Atomic.cuh:450]: It may be implemented and it may be useful, but we don't implement that in PyTorch at a kernel level. As mentioned above, all these orderings can often be simulated with the current API and a bit of imagination :)

Comment (ZelboK) [reply aten/src/ATen/cuda/Atomic.cuh:450]: I see - in that case I'll wait until @mikaylagawarecki has a chance to review again. Thanks for taking a look!

Comment (Franklalalala) [reply aten/src/ATen/cuda/Atomic.cuh:450]: @ZelboK I am working on [Tensor_network](https://en.wikipedia.org/wiki/Tensor_network), which require a series of matrix multiplication. In the case of complex elements, the torch scatter connot be used in GPU. As far as I concerned right now, we do not use sortage here, just elements' multiplication. 
By the way, we have worked a way out, that is, we transform the complex number through Euler transformation and turns the multiplication to addition of angles and multiplication of magnitude. 
The excellent work of you guys has reached out of my knowledge base, I connot give anymore advices. But thanks!

Comment (ZelboK) [reply aten/src/ATen/cuda/Atomic.cuh:450]: > @ZelboK I am working on [Tensor_network](https://en.wikipedia.org/wiki/Tensor_network), which require a series of matrix multiplication. In the case of complex elements, the torch scatter connot be used in GPU. As far as I concerned right now, we do not use sortage here, just elements' multiplication. By the way, we have worked a way out, that is, we transform the complex number through Euler transformation and turns the multiplication to addition of angles and multiplication of magnitude. The excellent work of you guys has reached out of my knowledge base, I connot give anymore advices. But thanks!

Thanks a lot for responding, I was genuinely curious. This helps give me perspective :)

**Pull request closed without merging**
